### DISCUSSION TITLE
type mutationCache.subscribe

### DISCUSSION THREAD

**Discussion by JoakimThorne** [TanStack/query#4151] [Q&A]
Hi,

I want to subscribe to the mutationCache to see when all mutations is completed. 
`mutationCache.subscribe((e: any) => {
        const { action, type, mutation } = e || {};
        console.log('action: ', action);
        console.log('type: ', type);
        console.log('mutation: ', mutation);
    });`

It seems like I cant type the `e` correctly. According to the docs it should be `MutationCacheNotifyEvent` but it does not seem to be exposed.

**Comment by TkDodo** [comment]
use type inference, `event` is automatically typed when doing:

```
mutationCache.subscribe((event) => {
})
```

see: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgWQK4wIY2BAdgYQwGMALAUzgF84AzKCEOAcgAFNcBnTIgawHooZYjAC0AR1RkoATyYAoOUTxc4IdFhwFi5OAF44uMgHcU67HkKkyACgCUCtZnNarAOg6oARhyJRgnm2syADcyXBhbPQA+REp7Pj44JOSU1LSkgD0AfiA

**Reply by JoakimThorne** [reply]
When logging the event I see an object called `action`. It dose not seem to be typed so I get errors when trying to access it. 

Im trying to do a global check when the mutations are done loading and enter state `success` and lookin on the object `event.action` it seems to hold the info I need

**Reply by TkDodo** [reply]
it is typed.

**Reply by JoakimThorne** [reply]
But In the link u sent and also in my env. I can only access `type` and `mutation`. So it seems lika `action` is not typed?

**Reply by JoakimThorne** [reply]
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgWQK4wIY2BAdgYQwGMALAUzgF84AzKCEOAcgAFNcBnTIgawHooZYjAC0AR1RkoATyYAoOUTxc4IdFhwFi5OAF44uMgHcU67HkKkyACgCUCtZnNarAOg6oARhyJRgnm2syADcyXBhbPQA+RAU4eLglTggAGzJXFIgAcyDQ8NcYaTAyewTE5VT0zJyQsJhXRw08UoSkjkqM7Ny612FNFvi2juru8Mi+PjgAHhEROHJBOABJRIxcOA4yCj68BUp7CbKj45O4AD0AfiA

**Reply by TkDodo** [reply]
The `MutationCacheNotifyEvent` is a union type, and not all events have an `action`:

https://github.com/TanStack/query/blob/7e51776bff00b2f6f8496a7a64b3a048238a4ee7/packages/query-core/src/mutationCache.ts#L30-L69

So TypeScript won't let you access `action` without checking for the `type` upfront.

**Reply by TkDodo** [reply]
> Im trying to do a global check when the mutations are done loading and enter state success and lookin on the object event.action it seems to hold the info I need

`useIsMutating` can do that

**Reply by JoakimThorne** [reply]
But `useIsMutating` only returns a number, I want to see errors and success as well

Want to be able to show "saving"/"saving done"/"error while saving"
